### ISSUE TITLE
Proposal: Better Strings.Replace function

### ISSUE THREAD

**Issue by Nukepayload2** [dotnet/vblang#390]
I found that the `Microsoft.VisualBasic.Strings.Replace` function had a strange behavior. 
The following program replaces Japanese full-width characters "ポ" and half-width characters "ﾎﾟ" with "po":
```vbnet
Module Program
    Sub Main(args As String())
        Dim replaced = Replace("ポポポポ",
                               Find:="ﾎﾟ", Replacement:="po",
                               Compare:=CompareMethod.Text)
        MsgBox(replaced)
    End Sub
End Module
```
The expected result is "popopopo". But actually the value of `replaced` is "popo". It's inconsistent with the `Range.Replace` function of Excel. Is this a bug?
If it's by design (for better performance or back-compatibility), I recommend adding another "Replace" function (Overloads the original `Replace` and optionally mark the original `Replace` obsolete) to help us reduce the number of bugs written in the future.

**Comment by VBAndCs** [comment]
This belongs to CorFx repo. You can repost it in https://github.com/dotnet/corefx/issues

**Comment by Nukepayload2** [comment]
@VBAndCs I don't think so. Because .NET Core 2.x does not support `Strings.Replace`. This repo is "The home for design of the Visual Basic .NET programming language and __runtime library__". Why can't I suggest a new `Function` for the VB runtime library here?

**Comment by reduckted** [comment]
> Why can't I suggest a new `Function` for the VB runtime library here?

You certainly can, but you seemed to be implying that the behavior is a bug. Bugs are better off being reported in the corefx repo. If it's determined that the function is behaving correctly, _then_ we can start discussing new functions.

> optionally mark the original `Replace` obsolete

As nice as that might be, it won't happen. The .NET team [try not to mark anything as obsolete](https://github.com/dotnet/corefx/issues/31181#issuecomment-407101473), as that would cause existing code to stop compiling if the `TreatWarningsAsErrors` option is enabled.

**Comment by AdamSpeight2008** [comment]
@Nukepayload2 They are treated as different characters, hence the result.

**Comment by VBAndCs** [comment]
@Nukepayload2
String.StartsWith, String.EndsWith, String.IndexOF and String.LastIndexOf have an overload that accepts StringComparison param, wher you can use the value  StringComparison.OrdinalIgnoreCase to prevent making any culture-specific transformation. String.Replace doesn't have such overload, so you can open an issue in CoreFx to ask for it.
Right now, you can write your own replace extenstion method that uses IndexOf(strToFind, StringComparison.OrdinalIgnoreCase) to solve your problem.

**Comment by Nukepayload2** [comment]
@AdamSpeight2008 
> They are treated as different characters
The `Replace` function treats "ポ" and "ﾎﾟ" are equal when finding the start index to replace. But when building the replaced string, the length of string segment to replace is incorrect. I've written a new sample to prove my guess:
```vbnet
Dim replaced = Replace(Expression:="ﾎﾟポ",
                       Find:="ポ", Replacement:="po",
                       Compare:=CompareMethod.Text)
' Guessed internal steps of `Replace`:
' 1. The Replace function uses CompareInfo.IndexOf to locate the start index to replace:
' startIndex = 0
' 2. The Replace function uses Find.Length to replace the first match.
' The result turned into poﾟポ
' 3. The Replace function uses CompareInfo.IndexOf to locate the next index to replace:
' startIndex = 2
' 4. The Replace function uses Find.Length to replace the second match.
' The result turned into poﾟpo
MsgBox(replaced) ' Output: poﾟpo
```

**Comment by Nukepayload2** [comment]
@VBAndCs 
Thanks for sharing your workaround. But I can't use the culture invariant compare mode. Because the users require that our product ignores the differences of full-width and half-width when finding and replacing data. I'm trying to implement my own `Replace` function to replace all `Strings.Replace`s in my code.

**Comment by jdmichel** [comment]
I see. The VB Replace function is completely different from the Replace method on the dotnet String object. It appears to have some logic to treat the ポ character as equal to the string "ﾎﾟ", but the problem is it's not quite right. 
I think you could call string.Replace  twice to get what you want, if you don't mind the inefficiency of two passes over the string. I would probably try that, and the only change it to a custom implementation if profiling shows a performance problem. 
`sourceText.Replace("ﾎﾟ", replacement).Replace("ポ"c, replacement)`

**Comment by KathleenDollard** [comment]
Since this looks like a bug, it would be very helpful to put an issue (copying from here) at https://github.com/dotnet/corefx. 

I don't mind it being here, but the folks that can fix it hang out in that repo and not here.
